lib/ansible/plugins/inventory/packet.py

```diff
@@ -0,0 +1,107 @@
+DOCUMENTATION = r'''
+    name: packet
+    plugin_type: inventory
+    author:
+      - Maxime Guyot (@Miouge1)
+    short_description: Ansible dynamic inventory plugin for Packet.
+    version_added: "2.9"
+    requirements:
+        - python >= 2.7
+        - packet
+    description:
+        - Reads inventories from the Packet API.
+        - Uses a YAML configuration file that ends with packet.(yml|yaml).
+        - The inventory groups are built from plan, tags and facility.
+    options:
+        plugin:
+            description: marks this as an instance of the 'packet' plugin
+            required: true
+            choices: ['packet']
+        auth_token:
+            description: The Packet API auth token
+            required: true
+            env:
+                - name: PACKET_AUTH_TOKEN
+        project_id:
+          description: Project ID
+          required: true
+'''
+
+EXAMPLES = r'''
+# Minimal example. `PACKET_AUTH_TOKEN` is exposed in environment.
+plugin: packet
+
+# Example with regions, types, groups and access token
+plugin: packet
+auth_token: foobar
+regions:
+  - eu-west
+types:
+  - g5-standard-2
+'''
+
+import os
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.module_utils.six import string_types
+from ansible.plugins.inventory import BaseInventoryPlugin
+
+
+try:
+    from packet import Manager
+except ImportError:
+    raise AnsibleError('the Packet dynamic inventory plugin requires packet-python.')
+
+
+class InventoryModule(BaseInventoryPlugin):
+
+    NAME = 'packet'
+
+    def verify_file(self, path):
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(('packet.yml', 'packet.yaml')):
+                return True
+        self.display.debug(
+            "Packet inventory filename must end with 'packet.yml' or 'packet.yaml'"
+        )
+        return False
+
+    def parse(self, inventory, loader, path, cache=True):
+        """Dynamically parse Packet the cloud inventory."""
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        self._read_config_data(path)
+
+        auth_token = self.get_option('auth_token')
+        if auth_token is None:
+            try:
+                auth_token = os.environ['PACKET_AUTH_TOKEN']
+            except KeyError:
+                pass
+
+        if auth_token is None:
+            raise AnsibleError((
+                'Could not retrieve Packet auth token '
+                'from plugin configuration or environment'
+            ))
+
+        manager = Manager(auth_token=auth_token)
+        params = {'per_page': 50}
+        devices = manager.list_devices(project_id=self.get_option('project_id'),
+                                       params=params)
+        for dev in devices:
+            host = dev.hostname
+            ip = dev.ip_addresses[0]['address']
+            for g in [dev.facility['code'], dev.plan['name']] + dev.tags:
+                self.inventory.add_group(g)
```

Comment:
instead of hardcoding the groups, i suggest adding the 'compose' functionality and allow users to determine how they create/name groups based on jinja2 expressions.

Reply:
Good idea. I've added the `compose` and `keyed_groups` in the latest commit